src/app/rpmostree-builtin-cancel.c

```diff
@@ -0,0 +1,122 @@
+                            GAsyncResult *res,
+                            gpointer      user_data)
+{
+  /* Nothing right now - we ideally should check for an error, but the
+   * transaction could have gone away. A better fix here would be to keep
+   * transactions around for some period of time.
+   */
+  (void) rpmostree_transaction_call_cancel_finish ((RPMOSTreeTransaction*)src, res, NULL);
+}
+
+static void
+on_active_txn_path_changed (GObject    *object,
+                            GParamSpec *pspec,
+                            gpointer    user_data)
+{
+  gboolean *donep = user_data;
+  *donep = TRUE;
+  g_main_context_wakeup (NULL);
+}
+
+int
+rpmostree_builtin_cancel (int             argc,
+                          char          **argv,
+                          RpmOstreeCommandInvocation *invocation,
+                          GCancellable   *cancellable,
+                          GError        **error)
+{
+  g_autoptr(GOptionContext) context = g_option_context_new ("");
+  glnx_unref_object RPMOSTreeSysroot *sysroot_proxy = NULL;
+  g_autofree char *transaction_address = NULL;
+  _cleanup_peer_ GPid peer_pid = 0;
+
+  if (!rpmostree_option_context_parse (context,
+                                       option_entries,
+                                       &argc, &argv,
+                                       invocation,
+                                       cancellable,
+                                       NULL, NULL,
+                                       &sysroot_proxy,
+                                       &peer_pid,
+                                       error))
+    return EXIT_FAILURE;
+
+
+  /* Keep track of the txn path we saw first, as well as checking if it changed
+   * over time.
+   */
+  g_autofree char *txn_path = NULL;
+  glnx_unref_object RPMOSTreeTransaction *txn_proxy = NULL;
+  if (!rpmostree_transaction_connect_active (sysroot_proxy, &txn_path, &txn_proxy,
+                                             cancellable, error))
+    return EXIT_FAILURE;
+  if (!txn_proxy)
+    {
+      /* Let's not make this an error; cancellation may race with completion.
+       * Perhaps in the future we could try to see whether a txn exited
+       * "recently" but eh.
+       */
+      g_print ("No active transaction.\n");
+      return EXIT_SUCCESS;
+    }
+
+  const char *title = rpmostree_transaction_get_title (txn_proxy);
+  g_print ("Cancelling transaction: %s\n", title);
+
+  /* Asynchronously cancel, waiting for the sysroot property to change */
+  rpmostree_transaction_call_cancel (txn_proxy, cancellable,
+                                     on_cancel_method_completed, NULL);
+
+  gboolean done = FALSE;
+  g_signal_connect (txn_proxy, "notify::active-transaction-path",
+                    G_CALLBACK (on_active_txn_path_changed), &done);
+  /* Wait for the transaction to go away */
+  while (!done)
+    {
+      const char *current_txn_path = rpmostree_sysroot_get_active_transaction_path (sysroot_proxy);
```

Comment:
We can just do this just once, no? E.g.:

1. register for transaction path changes
2. check that the path we're listening on still the same
3. enter main loop and wait until done

?

Reply:
I'm trying to be defensive against other things that might wake the mainloop.  I think it's a good pattern to use.